File: backend/app/schemas/permit.py
from __future__ import annotations

from datetime import date, datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field, field_validator

from app.core.validators import (
    MAX_DESCRIPTION_LENGTH,
    MAX_NAME_LENGTH,
    MAX_NOTES_LENGTH,
    CamelCaseModel,
    sanitize_string,
)
from app.schemas.user import UserResponse


# Permit Schemas
class PermitBase(BaseModel):
    permit_type: Literal[
        "building_permit",
        "occupancy_certificate",
        "completion_certificate",
        "environmental_permit",
        "fire_safety_approval",
    ]
    status: Literal[
        "not_applied", "applied", "under_review", "approved", "conditional", "rejected", "expired"
    ] = "not_applied"
    permit_number: str | None = Field(default=None, max_length=MAX_NAME_LENGTH)
    issuing_authority: str | None = Field(default=None, max_length=MAX_NAME_LENGTH)
    application_date: date | None = None
    approval_date: date | None = None
    expiration_date: date | None = None
    conditions: str | None = Field(default=None, max_length=MAX_DESCRIPTION_LENGTH)
    notes: str | None = Field(default=None, max_length=MAX_NOTES_LENGTH)

    @field_validator('permit_number', 'issuing_authority', 'conditions', 'notes', mode='before')
    @classmethod
    def sanitize_text(cls, v: str | None) -> str | None:
        return sanitize_string(v)


class PermitCreate(PermitBase):
    pass


class PermitUpdate(BaseModel):
    permit_type: Literal[
        "building_permit",
        "occupancy_certificate",
        "completion_certificate",
        "environmental_permit",
        "fire_safety_approval",
    ] | None = None
    status: Literal[
        "not_applied", "applied", "under_review", "approved", "conditional", "rejected", "expired"
    ] | None = None
    permit_number: str | None = Field(default=None, max_length=MAX_NAME_LENGTH)
    issuing_authority: str | None = Field(default=None, max_length=MAX_NAME_LENGTH)
    application_date: date | None = None
    approval_date: date | None = None
    expiration_date: date | None = None
    conditions: str | None = Field(default=None, max_length=MAX_DESCRIPTION_LENGTH)
    notes: str | None = Field(default=None, max_length=MAX_NOTES_LENGTH)

    @field_validator('permit_number', 'issuing_authority', 'conditions', 'notes', mode='before')
    @classmethod
    def sanitize_text(cls, v: str | None) -> str | None:
        return sanitize_string(v)


class PermitStatusUpdate(BaseModel):
    status: Literal[
        "not_applied", "applied", "under_review", "approved", "conditional", "rejected", "expired"
    ]


class PermitResponse(CamelCaseModel):
    id: UUID
    project_id: UUID
    permit_type: str
    status: str
    permit_number: str | None = None
    issuing_authority: str | None = None
    application_date: date | None = None
    approval_date: date | None = None
    expiration_date: date | None = None
    conditions: str | None = None
    notes: str | None = None
    created_at: datetime
    updated_at: datetime
    created_by: UserResponse | None = None


class PermitSummaryResponse(CamelCaseModel):
    id: UUID
    project_id: UUID
    permit_type: str
    status: str
    permit_number: str | None = None
    expiration_date: date | None = None
    created_at: datetime


class PermitComplianceReportResponse(CamelCaseModel):
    total_permits: int
    not_applied_count: int
    applied_count: int
    under_review_count: int
    approved_count: int
    conditional_count: int
    rejected_count: int
    expired_count: int
    expiring_soon_count: int
    permits_by_type: dict[str, int]

File: frontend/src/types/permit.ts
import { User } from './index'

// Permit Types
export type PermitType =
  | 'building_permit'         // heiter bniya
  | 'occupancy_certificate'   // tofes 4
  | 'completion_certificate'  // tofes 5
  | 'environmental_permit'
  | 'fire_safety_approval'

// Permit Status
export type PermitStatus =
  | 'not_applied'
  | 'applied'
  | 'under_review'
  | 'approved'
  | 'conditional'
  | 'rejected'
  | 'expired'

// Main Permit Interface
export interface Permit {
  id: string
  projectId: string
  permitType: PermitType
  status: PermitStatus
  permitNumber?: string
  issuingAuthority?: string
  applicationDate?: string
  approvalDate?: string
  expirationDate?: string
  conditions?: string
  notes?: string
  createdAt: string
  updatedAt: string
  createdBy?: User
}

// Permit Create Request
export interface PermitCreate {
  permitType: PermitType
  status?: PermitStatus
  permitNumber?: string
  issuingAuthority?: string
  applicationDate?: string
  approvalDate?: string
  expirationDate?: string
  conditions?: string
  notes?: string
}

// Permit Update Request
export interface PermitUpdate {
  permitType?: PermitType
  status?: PermitStatus
  permitNumber?: string
  issuingAuthority?: string
  applicationDate?: string
  approvalDate?: string
  expirationDate?: string
  conditions?: string
  notes?: string
}

// Permit Status Update Request
export interface PermitStatusUpdate {
  status: PermitStatus
}

// Permit Summary (compact version)
export interface PermitSummary {
  id: string
  projectId: string
  permitType: PermitType
  status: PermitStatus
  permitNumber?: string
  expirationDate?: string
  createdAt: string
}

// Permit Compliance Report
export interface PermitComplianceReport {
  totalPermits: number
  notAppliedCount: number
  appliedCount: number
  underReviewCount: number
  approvedCount: number
  conditionalCount: number
  rejectedCount: number
  expiredCount: number
  expiringSoonCount: number
  permitsByType: Record<string, number>
}

// Permit Filters
export interface PermitFilters {
  status?: PermitStatus[]
  permitType?: PermitType[]
  issuingAuthority?: string[]
  expiringWithinDays?: number
  searchQuery?: string
}

// Permit Sort Fields
export type PermitSortField =
  | 'permitType'
  | 'status'
  | 'permitNumber'
  | 'issuingAuthority'
  | 'applicationDate'
  | 'approvalDate'
  | 'expirationDate'
  | 'createdAt'
  | 'updatedAt'

// Permit Table Row
export interface PermitTableRow {
  id: string
  permitType: PermitType
  status: PermitStatus
  permitNumber?: string
  issuingAuthority?: string
  applicationDate?: string
  approvalDate?: string
  expirationDate?: string
  createdAt: string
  updatedAt: string
}

// Permit Column IDs
export type PermitColumnId =
  | 'permitType'
  | 'status'
  | 'permitNumber'
  | 'issuingAuthority'
  | 'applicationDate'
  | 'approvalDate'
  | 'expirationDate'
  | 'updatedAt'
